zeekclient/brokertypes.py

```diff
@@ -620,168 +620,193 @@ def from_broker(cls, data):
+        return [elem.to_py() for elem in self.elements]
 
     def to_broker(self):
         return {
             "@data-type": "vector",
-            "data": [elem.to_broker() for elem in self._elements],
+            "data": [elem.to_broker() for elem in self.elements],
         }
 
     @classmethod
     def from_broker(cls, data):
         res = Vector()
         for elem in data["data"]:
-            res._elements.append(from_broker(elem))
+            res.elements.append(from_broker(elem))
         return res
 
+    @property
+    def elements(self):
+        return self._elements
+
+    @elements.setter
+    def elements(self, elements):
+        if not isinstance(elements, tuple) and not isinstance(elements, list):
+            raise TypeError("Vector initialization requires tuple or list data")
+        if not all(isinstance(elem, Type) for elem in elements):
+            raise TypeError("Non-empty Vector construction requires brokertype values.")
+
+        self._elements = elements
+
 
 class Set(DataType):
     def __init__(self, elements=None):
-        self._elements = elements or set()
-        if not isinstance(self._elements, set):
-            raise TypeError("Set initialization requires set data")
-        if not all(isinstance(elem, Type) for elem in self._elements):
-            raise TypeError("Non-empty Set construction requires brokertype values.")
+        self.elements = elements or set()
 
     def __lt__(self, other):
         res = super().__lt__(other)
         if res != NotImplemented:
             return res
-        for el1, el2 in zip(sorted(self._elements), sorted(other._elements)):
+        for el1, el2 in zip(sorted(self.elements), sorted(other.elements)):
             if el1 < el2:
                 return True
-        if len(self._elements) < len(other._elements):
+        if len(self.elements) < len(other.elements):
             return True
         return False
 
     def __hash__(self):
-        return hash(tuple(sorted(self._elements)))
+        return hash(tuple(sorted(self.elements)))
 
     def __iter__(self):
-        return iter(self._elements)
+        return iter(self.elements)
 
     def __len__(self):
-        return len(self._elements)
+        return len(self.elements)
 
     def __contains__(self, key):
-        return key in self._elements
+        return key in self.elements
 
     def to_py(self):
-        return set(elem.to_py() for elem in self._elements)
+        return set(elem.to_py() for elem in self.elements)
 
     def to_broker(self):
         return {
             "@data-type": "set",
-            "data": [elem.to_broker() for elem in sorted(self._elements)],
+            "data": [elem.to_broker() for elem in sorted(self.elements)],
         }
 
     @classmethod
     def from_broker(cls, data):
         res = Set()
         for elem in data["data"]:
-            res._elements.add(from_broker(elem))
+            res.elements.add(from_broker(elem))
         return res
```

Comment:
Hmm, I'm not sure adding `@property` for elements is reasonable: That provides a "non-protected" interface to directly access and mutate the elements array.

Couldn't this `from_broker()` be rewritten instead as the following, just using the normal constructor passing in the `_elements` to use instead?
```
return Set({from_broker(elem) for elem in data["data"]})
```

Reply:
Same for vector and table, e.g.:
```
return Table({from_broker(elem["key"]): from_broker(elem["value"]) for elem in ...})
```

Reply:
TBH, I do not see how to do what you suggested. Would you mind showing me in more detail what you have in mind?

Reply:
Hope the below brokertypes.py change clarifies it. The black formating is a bit funky for the Table one, maybe an extra variable would look better. I'm not sure why the current code does what it does, I doubt there's a real performance argument to be made accessing _elements.

```
diff --git a/tests/test_brokertypes.py b/tests/test_brokertypes.py
index 70fb485..19d5a21 100755
--- a/tests/test_brokertypes.py
+++ b/tests/test_brokertypes.py
@@ -512,3 +512,25 @@ class TestBrokertypes(unittest.TestCase):
         data = b'{ "data": "foobar", "@data-type": "count" }'
         with self.assertRaisesRegex(TypeError, "invalid data for Count"):
             obj = Count.unserialize(data)
+
+    def test_container_from_broker(self):
+        s = Set.from_broker({"data": [{"@data-type": "string", "data": "s"}]})
+        self.assertEqual(1, len(s))
+
+        v = Vector.from_broker({"data": [{"@data-type": "string", "data": "s"}]})
+        self.assertEqual(1, len(v))
+
+        t = Table.from_broker(
+            {
+                "data": [
+                    {
+                        "key": { "@data-type": "string", "data": "s"},
+                        "value": {
+                            "@data-type": "integer",
+                            "data": "42",
+                        },
+                    }
+                ]
+            }
+        )
+        self.assertEqual(1, len(t))
diff --git a/zeekclient/brokertypes.py b/zeekclient/brokertypes.py
index 55b7bf8..094f243 100644
--- a/zeekclient/brokertypes.py
+++ b/zeekclient/brokertypes.py
@@ -666,10 +666,7 @@ class Vector(DataType):
 
     @classmethod
     def from_broker(cls, data):
-        res = Vector()
-        for elem in data["data"]:
-            res._elements.append(from_broker(elem))
-        return res
+        return Vector([from_broker(elem) for elem in data["data"]])
 
 
 class Set(DataType):
@@ -714,10 +711,7 @@ class Set(DataType):
 
     @classmethod
     def from_broker(cls, data):
-        res = Set()
-        for elem in data["data"]:
-            res._elements.add(from_broker(elem))
-        return res
+        return Set({from_broker(elem) for elem in data["data"]})
 
 
 class Table(DataType):
@@ -781,10 +775,12 @@ class Table(DataType):
 
     @classmethod
     def from_broker(cls, data):
-        res = Table()
-        for elem in data["data"]:
-            res._elements[from_broker(elem["key"])] = from_broker(elem["value"])
-        return res
+        return Table(
+            {
+                from_broker(elem["key"]): from_broker(elem["value"])
+                for elem in data["data"]
+            }
+        )

```

Reply:
Yeah I agree here, let's not make it a property. I like your alternative, Arne. But let me add two things for perspective:
- This module should soon leave `zeek-client` altogether and become part of the new Broker bindings. As-is, I consider all of it experimental. Once that migration happens we can and should have much more detailed discussion re. what the various APIs should look like.
- I don't think anything here was broken and all existing access to `_elements` was what you'd consider "friends" in C++. The single underscore is a convention, not a hard rule, so I don't think it's really worth much of our energy to debate. I'm shrugging at you, pylint...

Reply:
The patches now have the cleanups of the `from_broker` methods and the new test case as suggested by Arne. I locally disabled the diagnostics in `from_vec` and `from_py`.